### PR TITLE
Added parameters and exception behaviour to pqdm

### PR DESCRIPTION
**Pull request by Sherwin-14**
I did all the steps outlined in the issue #581, we can now build upon this.

----
📚 Documentation preview 📚: https://earthaccess--792.org.readthedocs.build/en/792/

### CODE REVIEW

**Comment by chuckwondo** [comment]
@mfisher87, I see lots of Ruff error messages regarding a number of files that @Sherwin-14 has not touched in this PR. Is there something already on `main` to address these errors, such that Sherwin simply needs to merge `main` into his own branch to silence them?

**Comment by mfisher87** [comment]
Not yet, but that should be an easy fix. I'll take care of it now :) https://github.com/nsidc/earthaccess/issues/788

**Comment by mfisher87** [comment]
https://github.com/nsidc/earthaccess/pull/793

**Comment by chuckwondo** [comment]
> Not yet, but that should be an easy fix. I'll take care of it now :) #788
> #793

Excellent! Thank you for the quick fix.

@Sherwin-14, Matt's fix is now merged into `main`, so please update your branch with the fix from `main`, which will eliminate the ruff errors in your build.

**Comment by chuckwondo** [comment]
@Sherwin-14, do you have enough information to proceed with the changes I requested?

**Comment by Sherwin-14** [comment]
> @Sherwin-14, do you have enough information to proceed with the changes I requested?

Yeah I have the info I needed, I had put this on hold since I was preoccupied with some other things for the last couple of days. I'll resume my work on this asap.

**Comment by betolink** [comment]
Would it be possible and in scope of this PR to use `**kwargs` and pass them to `pqdm` for both `open()` and `download()` with sane defaults? Users could be in control of the failing behavior but also the number of threads to use and the verbosity (don't display the progress bar). I'm not sure about the style, in Xarray they use `**kwargs` and an equivalent `backend_kwargs={}` that in our case could be something like `pqdm_opts={}`

```python
# code in api.py and something similar in store.py
# in this example we leave threads out to ensure backwards compatibility and will use it in the store class if present
def open(granules, treads, provider, pqdm_opts):
    default_pqdm_opts = {
      n_jobs=threads,
      exception_behaviour='immediate'
      colour="green",
      disable=False # True means no output including the progress bar.
    }

    if not pqdm_opts;
        pqdm_opts = default_pqdm_opts
    earthaccess.__store__.open(granules, treads, pqdm_opts)

```

and actually we should do the same with `fsspec` so we let users tune their caching behavior when opening the files, aka a **"smart_open"**. What do you all think?

**Comment by mfisher87** [comment]
> I had put this on hold since I was preoccupied with some other things for the last couple of days. I'll resume my work on this asap.

@Sherwin-14 No rush! We really appreciate any time you can spare, and don't want to ask for more than that :)

**Comment by mfisher87** [comment]
@betolink I really like that idea. Basically "Allow customizing all pqdm behaviors" (and separately, "Allow customizing all fsspec behaviors")?

I'm not sure how I feel about including that as part of this PR. I lean towards separate out of respect for @Sherwin-14 's time, but I don't want to speak for you, Sherwin!

**Comment by Sherwin-14** [comment]
> We decided during hack day today that we want to allow users to customize pqdm behaviors like so:
> 
> ```python
> def open(something, threads=16, *, pqdm_kwargs={...}, ...):
>     pqdm_kwargs_defaults = {"exception_behavior": "immediate", "threads": threads}
>     pqdm_kwargs = pqdm_kwargs_defaults.update(pqdm_kwargs)
>     ...
> ```
> 
> We considered the `threads` argument as fundamental to the function of earthaccess, so we think it should remain a top-level parameter. We'll use threads whether we use pqdm or not in 5 years.

How should we now proceed with the changes we discussed in the hackday?

**Comment by chuckwondo** [comment]
> > We decided during hack day today that we want to allow users to customize pqdm behaviors like so:
> > ```python
> > def open(something, threads=16, *, pqdm_kwargs={...}, ...):
> >     pqdm_kwargs_defaults = {"exception_behavior": "immediate", "threads": threads}
> >     pqdm_kwargs = pqdm_kwargs_defaults.update(pqdm_kwargs)
> >     ...
> > ```
> >   
> > We considered the `threads` argument as fundamental to the function of earthaccess, so we think it should remain a top-level parameter. We'll use threads whether we use pqdm or not in 5 years.
> 
> How should we now proceed with the changes we discussed during the hackday?

Do the same thing you did in the PR you already started for this, but instead of `fail_fast` (which is what you pretty much completed) change it to `pqdm_kwargs` and adjust your code changes similar to what is shown above, although the code above won't work because `pqdm` has an `n_jobs` arg, not a `threads` arg, so I suggest something like so, instead, within the functions/methods that invoke `pqdm`:

```python
pqdm_kwargs = {
    "exception_behavior": "immediate",
    "n_jobs": threads,
    **pqdm_kwargs,
}

...

pqdm(<other args>, **pqdm_kwargs)
```

**Comment by mfisher87** [comment]
Great point! Much clearer with the double splat too ;)

**Comment by chuckwondo** [comment]
@Sherwin-14, do you need some help with this?

**Comment by Sherwin-14** [comment]
> @Sherwin-14, do you need some help with this?

I will try to accomodate the change asap, I was caught up in something else up untill now

**Comment by chuckwondo** [comment]
> > @Sherwin-14, do you need some help with this?
> 
> I will try to accomodate the change asap, I was caught up in something else up untill now

No rush. I just wanted to check in to see if you need any help or if something else was requiring you to postpone your continuation.

**Comment by Sherwin-14** [comment]
@chuckwondo 

``` python
def download(
    granules: Union[DataGranule, List[DataGranule], str, List[str]],
    local_path: Optional[str],
    provider: Optional[str] = None,
    threads: int = 8,
    pqdm_kwargs: dict = {}
) -> List[str]:
    """Retrieves data granules from a remote storage system.

       * If we run this in the cloud, we will be using S3 to move data to `local_path`.
       * If we run it outside AWS (us-west-2 region) and the dataset is cloud hosted,
            we'll use HTTP links.

    Parameters:
        granules: a granule, list of granules, a granule link (HTTP), or a list of granule links (HTTP)
        local_path: local directory to store the remote data granules
        provider: if we download a list of URLs, we need to specify the provider.
        threads: parallel number of threads to use to download the files, adjust as necessary, default = 8

    Returns:
        List of downloaded files

    Raises:
        Exception: A file download failed.
    """
    provider = _normalize_location(provider)
    pqdm_kwargs = {
        "exception_behavior": "immediate",
        "n_jobs": threads,
        **pqdm_kwargs,
    }
    if isinstance(granules, DataGranule):
        granules = [granules]
    elif isinstance(granules, str):
        granules = [granules]
    try:
        results = earthaccess.__store__.get(
            granules, local_path, provider, threads, pqdm_kwargs
        )    
    except AttributeError as err:
        logger.error(
            f"{err}: You must call earthaccess.login() before you can download data"
        )
        return []
    return results

```

Just for clarification, the public API function download would look something like this. The `pqdm_kwargs` would be taken in at the user interface and then passed on to the lower level functions in the entire call chain isn't it? or am I missing something?

**Comment by chuckwondo** [comment]
Correct, but you _almost never_ want to use mutable values for default values, and this is one of those cases (among other references, see <https://docs.python-guide.org/writing/gotchas/#mutable-default-arguments>).

So, do NOT do this:

```python
def download(
    granules: Union[DataGranule, List[DataGranule], str, List[str]],
    local_path: Optional[str],
    provider: Optional[str] = None,
    threads: int = 8,
    pqdm_kwargs: dict = {}
) -> List[str]:
    pqdm_kwargs = {
        "exception_behavior": "immediate",
        "n_jobs": threads,
        **pqdm_kwargs,
    }
```

Instead, DO THIS:

```python
def download(
    granules: Union[DataGranule, List[DataGranule], str, List[str]],
    local_path: Optional[str],
    provider: Optional[str] = None,
    threads: int = 8,
    pqdm_kwargs: Optional[Mapping[str, Any]] = None,
) -> List[str]:
    pqdm_kwargs = {
        "exception_behavior": "immediate",
        "n_jobs": threads,
        **(pqdm_kwargs or {}),
    }
```

File: earthaccess/store.py
Comment (chuckwondo) [inline earthaccess/store.py:409]: Where did this come from?

File: earthaccess/api.py
Comment (chuckwondo) [inline earthaccess/api.py:208]: Please update the doctstring appropriately.

File: earthaccess/api.py
Comment (chuckwondo) [inline earthaccess/api.py:253]: Please update the docstring appropriately.

File: earthaccess/store.py
Comment (chuckwondo) [inline earthaccess/store.py:342]: Please update the docstring appropriately.

**Review by chuckwondo — CHANGES_REQUESTED** [review]
Thanks @Sherwin-14. In addition to my other comments, please add at least 1 unit test. I don't think you need to test passing all possible pqdm kwargs, but perhaps just test that you can produce an error for downloading and have it raise an error (not that the default exception behavior is "immediate") instead of returning with a list containing errors.

**Review by chuckwondo — CHANGES_REQUESTED** [review]
@Sherwin-14, I suspect you marked many suggestions as "resolved," but forgot to push the changes that resolve the suggestions.

Also, please remember to go back to the Conversations tab of your PR and, in the upper-right area of the tab, click the circular double-arrows next to a reviewer's name to request a re-review after the reviewer has requested changes. Otherwise, the reviewer doesn't know when you're finished making changes and ready for a follow-up review.

Clicking that icon will send another email indicating that you want another review.

image

**Comment by chuckwondo** [comment]
@Sherwin-14, several other PRs have landed on `main`, so please also be sure to merge the changes from `main` and resolve conflicts.

**Comment by Sherwin-14** [comment]
@chuckwondo Hey Chuck, I have made the appropirate changes but I haven't pushed the code yet since I was writing unittest for this one which you suggested earlier. I will re-request for a review once I feel this is ready to be reviewed

**Comment by chuckwondo** [comment]
> @chuckwondo Hey Chuck, I have made the appropriate changes but I haven't pushed the code yet since I was writing unittest for this one which you suggested earlier. I will re-request for a review once I feel this is ready to be reviewed

Sounds good, but please wait to mark any suggestions as resolved until you push the corresponding code changes, otherwise it is confusing to others.

**Comment by Sherwin-14** [comment]
```py
class TestEarthAccessDownload(unittest.TestCase):
    def test_download(self):
        # Call the login function
        earthaccess.login()

        # Call the search function
        results = earthaccess.search_data(
            short_name='ATL06',  # ATLAS/ICESat-2 L3A Land Ice Height
            bounding_box=(-10, 20, 10, 50),  # Only include files in area of interest...
            temporal=("1999-02", "2019-03"),  # ...and time period of interest.
            count=10
        )

        # Mock earthaccess.__store__
        with patch.object(earthaccess, '__store__', return_value= Mock()) as mock_store:
            with patch.object(mock_store, 'get') as mock_get:
                # Simulate a download failure
                mock_get.side_effect = Exception('Download failed')
    
                # Call the download function
                files = earthaccess.download(results, "/home/download-folder")
    
                # Assert that the download failed
                self.assertEqual(files, [])

if __name__ == '__main__':
    unittest.main()

```

This is the test function I came up with and this is giving me an error log that looks something like [this](https://0x0.st/XU3h.txt). Is this the thing we want to achieve at this point? Also if the test is correct then where should I place it should I make a completely new file for this like `test_api.py`?

**Comment by chuckwondo** [comment]
Hi @Sherwin-14. Thanks for working on the unit tests for this PR.

The first thing I suggest you do is merge the latest from `main` into your branch, so you're working with the latest code.

Regarding the test you've written, it looks like your generally heading in the right direction.

I have some suggestions:

- Sure, if you want to put your new tests in a new `tests/unit/test_api.py` file, that sounds good to me.
- Do not use the unittest library. There's no need for the additional "overhead". Simply use top-level test functions and built-in `assert` statements. We have almost completely moved away from unittest in our integration tests, and we have a mix in our unit tests, but we should continue moving away from unittest (although I suspect this may be up for debate).
- Regarding your test shown above, that looks good, but since the new behavior is for the call to `download` to fail immediately upon any single download failure (i.e., pqdm `exception_behavior="immediate"`), then you will want to put your call to `download` in a context manager, as by default we expect an exception to be raised (as shown in your linked test failure output), thus we do not expect to be able to assert that we get back an empty list of results:

```
        with patch.object(earthaccess, '__store__', return_value= Mock()) as mock_store:
            with patch.object(mock_store, 'get') as mock_get:
                mock_get.side_effect = Exception('Download failed')

                with pytest.raises(Exception, match="Download failed"):
                    earthaccess.download(results, "/home/download-folder")
```

**Comment by Sherwin-14** [comment]
I have taken out unittest but it seems that it is necessary to use unittest for importing the likes of Mock. Another option here would be to use the pytest mock library but that means we will need to add an additional dependency.

**Comment by chuckwondo** [comment]
> I have taken out unittest but it seems that it is necessary to use unittest for importing the likes of Mock. Another option here would be to use the pytest mock library but that means we will need to add an additional dependency.

You can still use unittest Mocks, but there's no need for creating a subclass of unittest.TestCase.

File: tests/unit/test_api.py
Comment (chuckwondo) [inline tests/unit/test_api.py:7]:
```suggestion
def test_download_immediate_failure(monkeypatch):
```

File: tests/unit/test_api.py
Comment (chuckwondo) [inline tests/unit/test_api.py:28]: Please also add a 2nd test like so (I haven't tested, so you may need to adjust it).

We want to make sure that overriding the default pqdm kwargs works. We should probably try a few more things, but this should at least suffice for now:

```python
def test_download_deferred_failure(monkeypatch):
    earthaccess.login()

    results = earthaccess.search_data(
            short_name='ATL06',  
            bounding_box=(-10, 20, 10, 50),  
            temporal=("1999-02", "2019-03"), 
            count=10,
    )

    def mock_get(*args, **kwargs):
        raise Exception('Download failed')

    mock_store = Mock()
    monkeypatch.setattr(earthaccess, '__store__', mock_store)
    monkeypatch.setattr(mock_store, 'get', mock_get)

    results = earthaccess.download(results, "/home/download-folder",
            {"exception_behavior": "deferred"})

    assert len(results) == 10 and all(isinstance(e, Exception), results)
```

**Review by chuckwondo — CHANGES_REQUESTED** [review]
In addition to my comments, please also make corrections to fix the pre-commit issues: https://results.pre-commit.ci/run/github/399867529/1730125212.zyU74SY_QNaQKq_PZmUrIQ

It looks like you perhaps have not installed pre-commit hooks locally, otherwise your attempt to commit your changes should have produced those same pre-commit issues locally, thus preventing your commit from completing.

Please install `pre-commit`, if you haven't already done so, an then run the following from your forked/cloned repo to install the hooks configured in the repo: `pre-commit install --install-hooks`. Then run `pre-commit run -a` to run the pre-commit checks against all the files, which should then help you fix the pre-commit errors.

**Review by chuckwondo — CHANGES_REQUESTED** [review]
It looks like you have not correctly merged/rebased from `main`. I see changes that I should not see. For example, in CHANGELOG.md, I see a new Fixed section, which is _not_ new. It is already on `main`, so should _not_ show up as something new on your branch.

I think there are some other spots that seem out of sync too, such as in the `earthacces/store.py` method `_open_files`, where the type of the `threads` parameter seems to have been clobbered. It should be `threads: int = 8` (a recent change), instead of `threads: Optional[int] = 8`.

**Review by chuckwondo — DISMISSED** [review]
Thanks @Sherwin-14.

I'm approving, but do @mfisher87, @betolink, or @jhkennedy want to take a pass over this before it gets merged?

File: tests/unit/test_api.py
Comment (mfisher87) [inline tests/unit/test_api.py:7]:
```suggestion
def test_download_immediate_failure(monkeypatch):
```

**Review by mfisher87 — COMMENTED** [review]
Looks amazing, @Sherwin-14 ! Thank you! A little nitpick, and let's merge this!

**Review by chuckwondo — APPROVED** [review]

**Pull request merged**
